Fix issue #28: Unintuitive adapter syntax

grzes314 opened the pull request.

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+        for adapted_name, recipe in self.adapter.items():
+            try:
+                adapted_steps[adapted_name] = self._adapt_one_name(step_inputs, recipe)
+            except (KeyError, ValueError) as ex:
```
@grzes314 I think the usual naming is simply e

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+
         return adapted_steps
 
     def unpack(self, step_inputs):
```
@grzes314 again I know it is on me but regardles I think unpack should be private

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
-        return unpacked_steps
+            unpacked_steps.update(step_dict)
+            for key in step_dict.keys():
+                key_to_step_names[key].append(step_name)
```
@grzes314 I have to say I use key_to_step_names.setdefault(key, []).append(step_name) instead of defaultdict(list) but it looks way cleaner here.

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+            return self._extract_one_item(step_inputs, input_name, key)
+
+        if isinstance(recipe, tuple) and len(recipe) == 2 and callable(recipe[1]):
+            lst = recipe[0]
```
@grzes314 since we assure that it is a tuple of len 2 then simply lst, fun = recipe will do

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+            lst = recipe
+            fun = lambda x: x
+        else:
+            raise ValueError("Invalid adapting recipe: '{}'".format(recipe))
```
@grzes314 we followed the msg = ''.format first then pass it to Exception/Error I would keep that convention

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+            try:
+                return input_dict[key]
+            except KeyError:
+                raise StepsError("Step '{}' didn't have '{}' in its output."
```
@grzes314 same first msg then pass it to StepsError

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+                raise StepsError("Step '{}' didn't have '{}' in its output."
+                                 .format(input_name, key))
+        except KeyError:
+            raise StepsError("Step '{}' doesn't have '{}' as its input step."
```
@grzes314 same first msg then pass it to StepsError

jakubczakon commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
+        return fun(extracted)
+
+    def _extract_one_item(self, step_inputs, input_name, key):
+        try:
```
@grzes314 maybe insead of try catching input_dict = step_inputs.get(input_name) and then if input_dict is not None else raise. I think there is a lot of try_catching here and maybe not all of it necessary. On the other hand user is breaching the contract when she passes incorrect step_inputs so it definitely should raise so why not try catch.

grzes314 replied:
This is a question of whether to use [EAFP](https://docs.python.org/3.6/glossary.html#term-eafp) style or [LBYL](https://docs.python.org/3.6/glossary.html#term-lbyl) style. I also find LBYL more readable, but AFAIK EAFP is preferred in Python.

jakubczakon commented on tests/steps_test_case.py:
```diff
@@ -0,0 +1,29 @@
+import logging
```
@grzes314 It is official we are testing! Happy as a clam.

jakubczakon commented on tests/test_base.py:
```diff
@@ -0,0 +1,133 @@
+import unittest
```
@grzes314 I am only thinking whether to use unittest or pytest. pytest is generally easier for non dev people  Do you feel strongly about unittest ?

grzes314 replied:
I think I chose unittest because of its similarity to JUnit which I used before. But I don't insist. I can rewrite these tests in pytest.

mromaniukcdl commented on tests/test_base.py:
```diff
@@ -0,0 +1,133 @@
+            name='test_adapter_recipe_with_list',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [],
```
What's this `'X': []` ?

grzes314 replied:
List of values in input steps to extract is empty, so `X` will be an empty list.

mromaniukcdl commented on tests/test_base.py:
```diff
@@ -0,0 +1,133 @@
+
+        self.assertEqual(res['X'], [])
+
+        self.assertTrue(np.array_equal(res['Y'][0], self.data['input_1']['features']))
```
Why does `res['Y']` have to be further indexed with `[0]`?

grzes314 replied:
`recipe` for `Y` is a list, so `res['Y']` is a list.

mromaniukcdl commented on tests/test_base.py:
```diff
@@ -0,0 +1,133 @@
+            name='test_adapter_recipe_with_list_and_function',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': ([('input_1', 'features'),
```
Personally I think this kind of adapter is hard to understand. My preference would be that if you want to concatenate features from different input steps then you should just create a concatenation step.

grzes314 replied:
First, let me notice that the syntax for this case hasn't changed :)

In this PR I executed my proposition described in issue #28 thread. My idea was to retain as much of existing behaviour as possible, while:
1. Simplifying syntax for the simplest and most popular case (removing brackets if you just want one item).
2. Making the syntax more consistent (recipe containing a list yields a list of objects).

If you have ideas for different approach to adapting inputs present them in issue #28 :)

mromaniukcdl commented on tests/test_base.py:
```diff
@@ -0,0 +1,133 @@
+            name='test_adapter_creates_defined_keys',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [('input_1', 'features')],
```
Wasn't the reason for making changes to get rid of the square brackets here?

grzes314 replied:
If you do `'X': ('input_1', 'features')`, then `X` will point to the same object as `features` in `input_1`.
If you do `'X': [('input_1', 'features')]`, then `X` will be a singleton list containing those `features`.

grzes314 commented on steps/base.py:
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
         for step_name, step_dict in step_inputs.items():
-            unpacked_steps = {**unpacked_steps, **step_dict}
-        return unpacked_steps
+            unpacked_steps.update(step_dict)
```
I check if this is a case a few lines further. If yes, an exception is thrown and `unpacked_steps` is discarded.

grzes314 commented:
@jakubczakon I took into account most of your suggestions. Most notably, I rewrote the tests in pytest.

grzes314 commented:
Following our recent discussion I introduced a few important changes in this PR:
- I moved adapting logic to a separate class.
- I removed option of adding a function to adapter.
- It is now possible to assign an arbitrary object to an adapted name.
- Unfortunately, it means that to let the adapter know we want to extract an element from the inputs we have to precede (input, key) pair with letter E, e.g. `'y': E('input_train', 'labels')`. If it was just `'y': ('input_train', 'labels')` there would be no telling if 'y' should be a constant tuple consisting of two strings, or if we should extract something from inputs results.

**The "adapter syntax" can now be (almost) any Python literal** and the adapter preserves structure of that literal's value with all `E(input, key)` replaced by elements extracted from the inputs.
